File: javascript/2 - FilaPrioridade.js

class Elemento {
  constructor(elemento, prioridade) {
    this.elemento = elemento;
    this.prioridade = prioridade;
  }
}

  
class FilaPrioridade {
  constructor() {
    this.fila = [];
  }
 
  adicionar(elemento, prioridade) {
    const elem = new Elemento(elemento, prioridade);
    let added = false

    for (let i = 0; i < this.fila.length; i++) {
      if ( elem.prioridade < this.fila[i].prioridade ) {
        this.fila.splice(i, 0, elem);
        added = true
        break
      }
    }

    if ( !added ) this.fila.push(elem)
  }

  remover() {
    return this.fila.shift();
  }

  vazia() {
    return this.fila.length === 0;
  }

  proximo() {
    return this.fila[0]
  }

  limpar() {
    this.fila = []
  }

  imprimir()  {
      console.log(this.fila)
  }
}

const fila = new FilaPrioridade();
fila.adicionar(1, 10)
fila.adicionar(1, 8)
fila.adicionar(1, 5)
fila.adicionar(1, 1)
fila.adicionar(1, 4)

console.log(fila.remover())
console.log(fila.remover())
console.log(fila.remover())
console.log(fila.remover())
console.log(fila.remover())

File: ruby/2 - fila_prioridade.rb
class Element
  attr_reader :value, :prioridade

  def initialize(value, prioridade)
    @value = value
    @prioridade = prioridade
  end

  def to_s
    "Valor: #{value} - Prioridade: #{prioridade}"
  end
end

class FilaPrioridade

  def initialize()
    @fila = []
  end

  def adicionar(value, prioridade)
    elm = Element.new(value, prioridade)
    added = false

    @fila.each_with_index do |element, index|
      if elm.prioridade < element.prioridade
        @fila.insert(index, elm)
        added = true
        break
      end
    end

    @fila.push(elm) unless added
  end

  def remover
    @fila.shift
  end

  def proximo
    @fila.first.value
  end

  def vazia?
    @fila.empty?
  end

  def limpar
    @file.clear
  end

  def tamanho
    @fila.size
  end

  def imprimir
    @fila.each do |element|
      puts element.to_s
    end
  end
end

fila = FilaPrioridade.new

fila.adicionar(1, 0)
fila.adicionar(2, 0)
fila.adicionar(100, 5)
fila.adicionar(101, 6)
fila.adicionar(99, 4)
fila.adicionar(10, 1)
puts fila.proximo
puts fila.tamanho
fila.imprimir
